pulp_service/pulp_service/tasking/new_worker.py

```diff
@@ -541,59 +541,65 @@ def fetch_task(self):
-                        task._locked_resources = exclusive_resources
-                        return task
-                    else:
-                        # Another worker claimed this task
-                        self._release_resource_locks(exclusive_resources)
+                # First try to acquire task lock (lightweight single Redis operation)
+                # This prevents wasting time on resource locks if another worker already has the task
+                task_lock_key = f"task:{task.pk}"
+                # Use SET with NX (only set if not exists) and EX (expiration in seconds)
+                # 24 hours = 86400 seconds
+                acquired = self.redis_conn.set(
+                    task_lock_key,
+                    self.name,
+                    nx=True,
+                    ex=86400
+                )
 
-                        if exclusive_resources:
-                            # This should NOT happen for tasks with exclusive resources!
-                            # We hold the resource locks, so no other worker should be able to claim this task
-                            raise RuntimeError(
-                                f"Worker {self.name} acquired resource locks {exclusive_resources} "
-                                f"for task {task.pk} but another worker claimed it with task lock. "
-                                f"This indicates a serious locking protocol violation."
+                if acquired:
+                    # Successfully claimed the task lock!
+                    # Now try to acquire resource locks for exclusive resources only
+                    # Shared resources are never locked
+                    if exclusive_resources:
+                        if self._try_acquire_resource_locks(exclusive_resources):
+                            # Got both task lock and resource locks!
+                            _logger.info(
+                                "Worker %s acquired task lock and resource locks for task %s in domain: %s",
+                                self.name,
+                                task.pk,
+                                task.pulp_domain.name
                             )
+                            # Store resources so we can release them later
+                            # Note: We don't store the task lock since it auto-expires in 24h
+                            task._locked_resources = exclusive_resources
+                            return task
                         else:
-                            # This is normal for tasks without exclusive resources where multiple workers
-                            # can acquire the (empty) resource locks simultaneously
+                            # Failed to acquire resource locks
+                            # Release the task lock since we can't execute this task
+                            self.redis_conn.delete(task_lock_key)
```

Comment:
I agree it's correct to delete the task_lock here as we acquired it and couldn't execute the task.